app/policies/action_policy.rb

```diff
@@ -0,0 +1,45 @@
+class ActionPolicy < ApplicationPolicy
+
+  ADMIN_OPERATIONS     = [Action::MEMO_OPERATION, Action::APPROVE_OPERATION, Action::DENY_OPERATION, Action::CANCEL_OPERATION].freeze
+  APPROVER_OPERATIONS  = [Action::MEMO_OPERATION, Action::APPROVE_OPERATION, Action::DENY_OPERATION].freeze
+  REQUESTER_OPERATIONS = [Action::CANCEL_OPERATION].freeze
+
+  class Scope < ApplicationPolicy::Scope
+    def resolve
```

Comment:
@hsong-rh How about this

```
def resolve
    my_scopes = @user.access_object.scopes('action','read')
    if my_scopes.include?('admin')
       scope.all
    elsif my_scopes.include?('group')
       action_ids = approver_id_list(scope.model.table_name)
       Rails.logger.debug { "Approver scope for actions: #{action_ids}" }
       scope.where(:id => action_ids)
    elsif my_scopes.include?('user')
       # What should the user see is that by owner
    else
       raise Exceptions::NotAuthorizedError, "Read access not authorized for #{scope}"
    end
end
```

Reply:
👍

Reply:
@mkanoor I prefer to put class level `permission_check` in the beginning. There's no need to proceed further if it fails.

Reply:
@hsong-rh permission_check is not using the scopes. Its just checking if you have the permission. The scopes method allows you do the 2 things
(1) Check permission
(2) Get Scope
Versus
scopes in a single pass collect all the scopes. If the scopes result is empty you dont have access

Reply:
Yes, I know the difference. My point is put `permission_check` in the beginning. If it fails, there's no need to continue. `scope` check is the second step, like what we define here.

Reply:
@hsong-rh So what happens if you have access, then you call the function again it has to loop thru all the permission one more time to pick up the scope. You are optimizing for what happens if there is no access, but if there is access you have to traverse the list again to collect scopes.

Reply:
No, `permission_check` only runs one time and it only checks permission, never list resources.